### DISCUSSION TITLE
A new term in Navier Stokes equation coupled with Cahn Hilliard equation

### DISCUSSION THREAD

**Discussion by Wolke926** [idaholab/moose#24993] [Q&A Modules: Navier-Stokes]
Dear MOOSE experts,

I want to write a .C file for my own equation for a new term, this is the one I want to add to the Navier-Stokes equation, phi is the order parameter in Cahn Hilliard equation, and the variational derivative of energy function respect phi is chemical potential w.
![Screen Shot 2023-07-19 at 8 55 11 PM](https://github.com/idaholab/moose/assets/120539827/c484ba74-e6d5-4dbb-b9c8-147a1f50079e)
    In the NS equation, the variable we want to solve is the velocity, so the w(chemical potential) and phi(order parameter) are coupled variables of kernel "SplitCHWRes" and kernel "SplitCHParsed" respectively, so how to add these variables in params and link these kernels in the Constructor initialization list?

![Screen Shot 2023-07-19 at 9 21 55 PM](https://github.com/idaholab/moose/assets/120539827/0ba4575e-9efb-45d6-868b-63e8e869e47f)

![Screen Shot 2023-07-19 at 9 26 40 PM](https://github.com/idaholab/moose/assets/120539827/077fc838-740c-4950-91f8-11327c813482)

There is the draft code, I marked the uncertain part, can anyone give me some suggestions?

#include "SurfaceTension.h"

registerMooseObject("NavierStokesApp", SurfaceTension);

InputParameters
SurfaceTension::validParams()
{
  InputParameters params = ADVectorKernelValue ::validParams();
  params.addClassDescription("Add surface tension term in NS equation");
  params.addCoupledVar("u", "velocity"); // but the u should be a vetor
  
   params.addCoupledVar("w", "chemical potential");   // ?
   params.addCoupledVar("phi", "order parameter");    // ?
  
  return params;
}

SurfaceTension::SurfaceTension(const InputParameters & parameters)
  : ADVectorKernelvalue(parameters),
    phi(),  //???? the variable in C-H equation 
    _grad_w ()     // ?? w is the chemical potential, namely the variable in SplitChWREs
{
}

ADRealVectorValue
SurfaceTension::precomputeQpResidual()
{
  
  return  _phi [_qp] * _grad_w[_qp] * _test[_i][_qp];
}

**Comment by GiudGiud** [comment]
Hello

It should be something like:
_phi(adCoupledValue("phi"),
_grad_w(adCoupledGradient("w")

in the initializer s list

and in the header:
const VariableValue & _phi;
const VariableGradient & _grad_w;

Guillaume

**Reply by Wolke926** [reply]
Thanks for your suggestions.
But I can not find  "adCoupledValue" but "CoupledValue" like in the file SoretDiffusion.C. 
Here is the .C and .h file. Are there other headers I should include?

![Screenshot 2023-07-24 at 9 30 12 PM](https://github.com/idaholab/moose/assets/120539827/b274b434-c709-4d58-9834-7db298b63ca9)
![Screenshot 2023-07-24 at 9 31 46 PM](https://github.com/idaholab/moose/assets/120539827/c052ba40-fff6-4b63-a784-9b856a244f47)

**Reply by GiudGiud** [reply]
In the header it should be ADVectorKernelValue not KernelADKernelValue

it s spelled coupledValue or coupledGradient no capitals at the beginning
For the AD version it s 
adCoupledValue and adCoupledGradient

**Reply by Wolke926** [reply]
Where is the file ADVectorKernelValue?
![Screenshot 2023-07-25 at 1 15 34 PM](https://github.com/idaholab/moose/assets/120539827/9fa6230d-158f-4fc6-ace4-76ee421d2596)

**Reply by GiudGiud** [reply]
We probably got the name not quite right. Search the kernels folder for it

**Reply by GiudGiud** [reply]
Ok I had some time to check and:

- ADVectorKernelValue does NOT exist. ADKernelValue is a simplified kernel base class that multiplies the "pre"residual generated by `precomputeQpResidual` by the test function. And we do NOT have the vector version of this in the framework. so for now use: `ADVectorKernel`. It's actually defined in `ADKernel.h` if you want to take a look. 

To use that base class, you will need to implement `computeQpResidual` in your derived class instead of `precomputeQpResidual`

**Reply by Wolke926** [reply]
Thanks! @GiudGiud

**Comment by Wolke926** [comment]
Sorry, I still got some problems when I compile them, what's the issue in my code?
![Screenshot 2023-08-01 at 1 10 50 PM](https://github.com/idaholab/moose/assets/120539827/7b762e5f-2267-4b0a-8416-943650a4821a)
![Screenshot 2023-08-01 at 1 10 35 PM](https://github.com/idaholab/moose/assets/120539827/ac210bf7-24c1-4a30-8e4c-0508403b7738)
![Screenshot 2023-08-01 at 1 10 26 PM](https://github.com/idaholab/moose/assets/120539827/42307310-02cb-4810-9e94-d3ba9ef9f02b)

**Reply by GiudGiud** [reply]
ComputeQpResidual should return ADReal not ADRealVectorValue

**Reply by GiudGiud** [reply]
Also you need the correct return type (ADReal) for the declaration of that routine in the header. You have it wrong there with the class name instead of a return type

**Reply by Wolke926** [reply]
I modified the code but there are still some errors. Do you think the return type is a vector? Why it shows the "redefinition of  ADKernelTempl "?
<img width="1164" alt="Screenshot 2023-08-02 at 8 44 30 PM" src="https://github.com/idaholab/moose/assets/120539827/82db6b89-3604-4fd7-b966-2732a8d77ead">
<img width="546" alt="Screenshot 2023-08-02 at 8 44 45 PM" src="https://github.com/idaholab/moose/assets/120539827/8bf26362-68f8-4172-aa7b-be6aff44d1a5">
<img width="451" alt="Screenshot 2023-08-02 at 8 45 03 PM" src="https://github.com/idaholab/moose/assets/120539827/7ced44bd-3541-482e-8d11-60ba21a0f154">

**Reply by GiudGiud** [reply]
Thé déclaration of the constructor in the source file is wrong. Please look at other kernels for examples

**Reply by Wolke926** [reply]
and what does this error mean? 
" no viable conversion from 'const MooseArray<ADReal>' to 'const MooseArray<libMesh::Real>'
    _phi(adCoupledValue ("phi")),  "

**Reply by GiudGiud** [reply]
SurfaceTension::SurfaceTension(parameters),
  ADReal(parameters)
...

this is wrong.
Please dont post screenshots of code, next time simply copy-paste

**Reply by Wolke926** [reply]
Sorry GiudGiud, I didn't realize the screenshot is less convenient.
I compared the file of "ADThermoDiffusion" with mine and I modified the "const VariableValue" to "const ADVariableValue" in the header file, the error seems to be reduced.
<img width="1260" alt="Screenshot 2023-08-03 at 4 56 42 PM" src="https://github.com/idaholab/moose/assets/120539827/ba522930-8152-482f-89cc-5adcc8d72d63">

***************************
#pragma once
 
#include "ADKernel.h"   

class SurfaceTension : public ADKernel 
{
public:
  static InputParameters validParams();

  SurfaceTension(const InputParameters & parameters);

protected:

 virtual ADReal computeQpResidual() override;

 const ADVariableValue & _phi;
 const ADVariableGradient & _grad_w;

};
***************************************************
#include "SurfaceTension.h"

registerMooseObject("cxyApp", SurfaceTension);

InputParameters
SurfaceTension::validParams()
{
  InputParameters params = ADKernel::validParams();
  params.addClassDescription("Add surface tension term in NS equation");
  params.addCoupledVar("u", "velocity"); 
  params.addCoupledVar("w", "the coupled chemical potential");   
  params.addCoupledVar("phi", "the coupled order parameter");    
  
  return params;
}

SurfaceTension::SurfaceTension(const InputParameters & parameters)
  : ADKernel(parameters),
    _phi(adCoupledValue ("phi")),        
    _grad_w(adCoupledGradient ("w"))    
{
}

ADReal   
SurfaceTension::computeQpResidual()
{
  return  _phi[_qp] * _grad_w[_qp] * _test[_i][_qp];
}

**Reply by GiudGiud** [reply]
You re missing an index. You have to index into the component of grad_w maybe?
The quantity returned is a vector, and you need to be returning a single ADReal for the residual
